lib/change_stream.js

```diff
@@ -476,96 +475,73 @@ function waitForTopologyConnected(topology, options, callback) {
   }, 3000); // this is an arbitrary wait time to allow SDAM to transition
 }
 
-// Handle new change events. This method brings together the routes from the callback, event emitter, and promise ways of using ChangeStream.
-function processNewChange(args) {
-  const changeStream = args.changeStream;
-  const error = args.error;
-  const change = args.change;
-  const callback = args.callback;
-  const eventEmitter = args.eventEmitter || false;
-  const cursor = changeStream.cursor;
-
-  // If the cursor is null, then it should not process a change.
-  if (cursor == null) {
-    // We do not error in the eventEmitter case.
-    if (eventEmitter) {
-      return;
-    }
+// Wait for topology to re-connect, then create a new cursor.
+// If in eventEmitter mode, returns null and doesn't iterate the change stream.
+function resumeFromError({ changeStream, callback, eventEmitter }) {
+  const { cursor } = changeStream;
+  const {
+    options: { readPreference },
+    resumeOptions
+  } = cursor;
 
-    const error = new MongoError('ChangeStream is closed');
-    return typeof callback === 'function'
-      ? callback(error, null)
-      : changeStream.promiseLibrary.reject(error);
-  }
+  // stop listening to all events from old cursor
+  ['data', 'close', 'end', 'error'].forEach(event => cursor.removeAllListeners(event));
 
-  const topology = changeStream.topology;
-  const options = changeStream.cursor.options;
-  const wireVersion = maxWireVersion(cursor.server);
+  // close internal cursor, ignore errors
+  cursor.close();
 
-  if (error) {
-    if (isResumableError(error, wireVersion) && !changeStream.attemptingResume) {
-      changeStream.attemptingResume = true;
-
-      // stop listening to all events from old cursor
-      ['data', 'close', 'end', 'error'].forEach(event =>
-        changeStream.cursor.removeAllListeners(event)
-      );
+  // throw it away
+  changeStream.cursor = null;
 
-      // close internal cursor, ignore errors
-      changeStream.cursor.close();
+  // attempt resume
+  const resumeAttempt = maybePromise(callback, cb => {
+    waitForTopologyConnected(changeStream.topology, { readPreference }, err => {
+      if (err) {
+        if (eventEmitter) {
+          changeStream.emit('error', err);
+          changeStream.emit('close');
+        }
+        return cb(err, null);
+      }
+      createChangeStreamCursor(changeStream, resumeOptions);
 
-      // attempt recreating the cursor
       if (eventEmitter) {
-        waitForTopologyConnected(topology, { readPreference: options.readPreference }, err => {
-          if (err) {
-            changeStream.emit('error', err);
-            changeStream.emit('close');
-            return;
-          }
-          changeStream.cursor = createChangeStreamCursor(changeStream, cursor.resumeOptions);
-        });
-
-        return;
+        cb();
+      } else {
+        changeStream.next(cb);
```

Comment:
I think perhaps `resumeFromError` is doing a bit too much and the readabilty suffers for it (it seems more like `resumeFromErrorAndCallNext` now). While there was a lot of duplication in the old code, I could still read it and see that we were:
  - waiting for the topology to become connected again (not a concern in `master` to remind you)
  - emitting an error if it existed
  - creating a new cursor
  - calling next on it

Now that is just called "resumeFromError" which happens to return a call to `next` on the cursor - feels a bit magical. 

It seems the primary benefit of this refactoring is to reduce the duplication of calls to `waitForTopologyConnected` and recreating the cursor. Maybe it's enough to do something like:
```
function resumeFromError(..., callback) {
   waitForConnected(..., () => {
     // replace cursor
     callback(undefined, cursor);
   });
}

cons result = maybePromise(.., cb => {
  resumeFromError(..., (err, cursor) => {
    if (err) return cb(err);
    cursor.next(cb);
  });
});
```